internal/sftp/goclient_test.go

```diff
@@ -0,0 +1,252 @@
+package sftp_test
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gliderlabs/ssh"
+	gosftp "github.com/pkg/sftp"
+	gossh "golang.org/x/crypto/ssh"
+	"gotest.tools/v3/assert"
+	tfs "gotest.tools/v3/fs"
+
+	"github.com/artefactual-sdps/enduro/internal/sftp"
+)
+
+// PubkeyHandler handles checking the client's public key against the keys in
+// the authorized_keys file.
+func pubkeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
+	file, err := os.Open("./testdata/authorized_keys")
+	if err != nil {
+		log.Fatalln("SSH: couldn't open authorized_keys file.")
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		allowed, _, _, _, err := ssh.ParseAuthorizedKey([]byte(scanner.Text()))
+		if err != nil {
+			log.Fatalln("SSH: couldn't parse authorized key.")
+		}
+		if ssh.KeysEqual(key, allowed) {
+			return true
+		}
+	}
+
+	log.Println("SSH: unknown key provided.")
+	return false
+}
+
+// HostKeySigner signs messages from the server to the client and allows the
+// client to confirm the host key signature.
+func hostKeySigner() (gossh.Signer, error) {
+	keyfile := "./testdata/serverkeys/test_rsa"
+
+	key, err := os.ReadFile(keyfile)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't read keyfile %q, %v\n", keyfile, err)
```

Comment:
Remember you can accept *t.Testing and use t.Fatal etc everywhere in the context of a test.

Reply:
What's the advantage of passing `*t.Testing` into setup functions that are not actually running tests, rather than just using the standard Go error return idiom?  If `t.Testing` isn't required in the function other than for stopping the tests, then it seems like an unnecessary argument to pass around.

Reply:
I think that t.Testing expects us to lower the bar when it comes to flow control in order to simplify our tests, e.g. we can write helpers that save the caller from looking up the errors:

```go
key, err := os.ReadFile(keyFile)
assert.NilError(t, err, "couldn't read keyfile %q", keyfile)
```

Or we can use `t.Cleanup` to save the caller from cleaning up resources, e.g. startSFTPServer

```go
func startSFTPServer(addr string) (*ssh.Server, error) { ... }

func TestGoClient(t *testing.T) {
	// ...
	srv, err := startSFTPServer(serverAddress)
	if err != nil {
		t.Fatalf("Couldn't start SFTP server: %v", err)
	}
	t.Cleanup(func() { srv.Close() })
	// ...
}

Could be simplified if accepted `*t.Testing`:

```go
func startSFTPServer(t *testing.T, addr string) *ssh.Server) {
	// srv := ...
	t.Cleanup(func() { srv.Close() })  
}

func TestGoClient(t *testing.T) {
	// Create SFTP server
	srv := startSFTPServer(serverAddress)
	// ...
}
```

Reply:
> ```
> key, err := os.ReadFile(keyFile)
> assert.NilError(t, err, "couldn't read keyfile %q", keyfile)
> ```

It seems weird to me to use `assert` for code that is not actually part of the tests.  The SFTP server is an unfortunate test set up requirement in my opinion, if I could test the client code without it I'd be much happier.  I also like that I could theoretically copy and paste the code as outside of the test context and it would run without having to remove the testing library dependencies.

Reply:
I understand, and it's valid to optimize for code reuse outside the testing context, even if it makes your tests a bit heavier.

Reply:
> ```
> func startSFTPServer(t *testing.T, addr string) *ssh.Server) {
>	 // srv := ...
>	 t.Cleanup(func() { srv.Close() })  
> }
> ```

Setting up the `t.Cleanup()` inside of `startSFTPServer` makes sense to me to make it easier to add more tests without having to worry about teardown each time.

Reply:
I don't know if there's a better way to break out of the `srvwait` loop then what I've done.  I could move the wait loop to a separate function, but I'm not sure if that's any better. :shrug:

Reply:
I like func() better. :)